File: web-frontend/src/app/core/user/services/user.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { CreateUserDto, UpdateUserDto, UserDataDto, UserSearchDto } from "../models/User";
import { Observable } from "rxjs";
import { PaginatedResults, UserSearchParams } from "../../../shared/search/models/Search";
import { SearchUrlBuilderService } from "../../../shared/search/services/SearchUrlBuilderService";
import { Language } from "../../../shared/language/models/Language";

@Injectable({
    providedIn: 'root'
})
export class UserService {
    private apiUrl: string = "http://localhost:8080/api/v1/users";

    constructor(
        private readonly urlBuilderService: SearchUrlBuilderService,
        private readonly http: HttpClient
    ) {}

    getUserByUsername(username: string): Observable<UserDataDto> {
        return this.http.get<UserDataDto>(`${this.apiUrl}/${username}`);
    }

    searchUsers(searchParams: UserSearchParams): Observable<PaginatedResults<UserSearchDto>> {
        const url = this.urlBuilderService.buildUserSearchUrl(this.apiUrl + "/search", searchParams);

        return this.http.get<PaginatedResults<UserSearchDto>>(url);
    }

    createUser(userDto: CreateUserDto): Observable<UserDataDto> {
        return this.http.post<UserDataDto>(this.apiUrl, userDto);
    }

    updateUser(userDto: UpdateUserDto): Observable<UserDataDto> {
        return this.http.put<UserDataDto>(this.apiUrl, userDto);
    }

    deleteUser(id: number) {
        this.http.delete(`${this.apiUrl}/${id}`);
    }

}

File: backend/hilbert-main/src/main/java/com/hilbert/core/user/service/UserService.java
package com.hilbert.core.user.service;

import com.hilbert.core.user.dto.*;
import com.hilbert.shared.search.models.PaginatedResults;
import com.hilbert.shared.search.models.SearchParams;
import com.hilbert.shared.search.models.UserSearchParams;

import java.util.List;

public interface UserService {

    UserDataDto getByUsername(String username);
    List<UserSmallDto> getByIds(List<Long> ids);
    PaginatedResults<UserSearchDto> searchUsers(UserSearchParams searchParams);
    UserDataDto createUser(CreateUserDto userDto);
    UserDataDto updateUser(UpdateUserDto userDto);
    void deleteUser(Long id);
}
